File: GFG/Remove loop in Linked List.java

/*Please note that it's Function problem i.e.
you need to write your solution in the form of Function(s) only.
Driver Code to call/invoke your function is mentioned above.*/

/*Complete The function
Node is as follows:
class Node {
    int data;
    Node next;
    Node(int d) {
        data = d;
        next = null;
    }
}*/
class GfG{
	  int removeTheLoop(Node node) {
	      Node slow = node;
	      Node fast = node;
	      while(slow != null && fast != null && fast.next != null){
	          slow = slow.next;
	          fast = fast.next.next;
	          if(slow == fast){
	              remove(slow, node);
	          }
	      }
	      return 1;
    }
    void remove(Node ptr, Node head){
        Node prev = null;
        while(ptr != head){
            prev = ptr;
            ptr = ptr.next;
            head = head.next;
        }
        prev.next = null;
    }
}

File: GFG/Remove loop in Linked List.cpp
/*The structure of linked list is the following
struct Node
{
int data;
Node* next; 
};*/

/*The function removes the loop from the linked list if present
You are required to complete this method*/
void removeTheLoop(Node *node)
{
     //Your code here
    Node *ptr1 = node, *ptr2 = node;
    if(ptr1 == NULL)return;
    
    while(ptr1 && ptr2 && ptr2->next){
        ptr1 = ptr1 -> next;
        ptr2 = ptr2 -> next -> next;
        if(ptr1 == ptr2){
            ptr1 = node;
            Node* temp = ptr2;
            while(ptr1 != ptr2){
                temp = ptr2;
                ptr1 = ptr1 -> next;
                ptr2 = ptr2 -> next;
            }
            temp -> next = NULL;
            break;
        }
    }
    
}
